### PR TITLE
:tada: New Destination: Vectara

### PR DESCRIPTION
**Pull request by ofermend**
* This PR is based on the wonderful work started in https://github.com/airbytehq/airbyte/pull/31958 
* Added parallelization via threads, icon, and a few minor robustness updates.

CC @flash1293

### CODE REVIEW

File: airbyte-integrations/connectors/destination-vectara/destination_vectara/client.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            delete_document_response = self._request(
+                endpoint="delete-doc",
+                data={
+                    "customerId": self.customer_id, 
+                    "corpusId": self.corpus_id,
+                    "documentId": document_id
+                    }
+                )
+            # TODO whether this is needed?
+            if delete_document_response:
```
Comment (flash1293) [inline airbyte-integrations/connectors/destination-vectara/destination_vectara/client.py:171]: This stuff doesn't seem to be necessary

File: docs/integrations/destinations/vectara.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+### Setup the Chroma Destination in Airbyte
+
+You should now have all the requirements needed to configure Chroma as a destination in the UI. You'll need the following information to configure the Chroma destination:
+
+- (Required) OAuth2.0 Credentials
+  - (Required) **Client ID**
+  - (Required) **Client Secret**
+- (Required) **Customer ID**
+- (Required) **Corpus Name**
```
Comment (flash1293) [inline docs/integrations/destinations/vectara.md:43]: We should document how the metadata and text field configurations work / what to put in there

File: airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            self.ids_to_delete.append(document_id)
+        else:
+            document_id = str(uuid.uuid4().int)
+
+        self.write_buffer.append((document_section, document_metadata, document_id))
+        if len(self.write_buffer) == self.flush_interval:
+            self.flush()
+
+    def flush(self) -> None:
+        """Writes to Convex"""
```
Comment (flash1293) [inline airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py:62]: Comment is outdated

File: airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        """Adds messages to the write queue and flushes if the buffer is full"""
+
+        stream_identifier = self._get_stream_id(record=record)
+        document_section = self._get_document_section(record=record)
+        document_metadata = self._get_document_metadata(record=record)
+        primary_key = self._get_record_primary_key(record=record)
+
+        if primary_key:
+            document_id = f"Stream_{stream_identifier}_Key_{primary_key}"
+            self.ids_to_delete.append(document_id)
```
Comment (flash1293) [inline airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py:53]: It should probably only add that to that array if the sync mode of the stream is actually append_dedupe, right?

**Review by flash1293 — COMMENTED** [review]
Thanks for working through it @ofermend ! There is one bug and some smaller nits I found; I will take a deeper look tomorrow

File: airbyte-integrations/connectors/destination-vectara/destination_vectara/client.py (hunk 1 above)
Comment (ofermend) [reply airbyte-integrations/connectors/destination-vectara/destination_vectara/client.py:171]: Yes agree. Was from original code. Will remove

File: airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    def __init__(self, client: VectaraClient, text_fields: Optional[List[str]], metadata_fields: Optional[List[str]], catalog: ConfiguredAirbyteCatalog):
+        self.client = client
+        self.text_fields = text_fields
+        self.metadata_fields = metadata_fields
+        self.streams = {f"{stream.stream.namespace}_{stream.stream.name}": stream for stream in catalog.streams}
+        self.ids_to_delete: List[str] = []
+
+    def delete_streams_to_overwrite(self, catalog: ConfiguredAirbyteCatalog) -> None:
+        streams_to_overwrite = [
+            stream.stream.name for stream in catalog.streams if stream.destination_sync_mode == DestinationSyncMode.overwrite
```
[the remark this answers is not in the document]
Comment (ofermend) [reply airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py:34]: yes good point. 
When we have a record that function uses f"{record.namespace}_{record.stream}"
What would be the equivalent in a stream schema? I see stream.stream.name which I assume is the second part. What should we use for the namespace?

[the remark this answers is not in the document]
Comment (flash1293) [reply airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py:34]: It’s stream.stream.namespace

File: airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py (hunk 4 above)
Comment (ofermend) [reply airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py:53]: Yes. Is the value "append_dedup" though (no "e")?

**Comment by ofermend** [comment]
Thanks @flash1293 - made some changes based on your comments but have a few more to do tmrw.

File: airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py (hunk 4 above)
Comment (ofermend) [reply airbyte-integrations/connectors/destination-vectara/destination_vectara/writer.py]: See my fix - is that how it should be checked?

File: airbyte-integrations/connectors/destination-vectara/unit_tests/test_destination.py
[the remark this answers is not in the document]
Comment (ofermend) [reply airbyte-integrations/connectors/destination-vectara/unit_tests/test_destination.py:1]: It's missing ConfigModel - can you please explain what ConfigModel needs to be and how it works?

**Comment by flash1293** [comment]
Thanks for the adjustments @ofermend 

I just pushed the following things:
* Square and pad the icon properly - size is great now!
* Add integration tests which actually interact with vectara (I created a test account for this) - this seemed like the best way of ensuring things work end to end
* Format code
* Fix the lookup for the stream whether it's in dedupe mode or not (what was necessary is to look it up from the catalog)

Otherwise this looks pretty good now, the only thing I'm not sure about is the parallel sending of data - will we run into issues with overwhelming the vectara service and it will start rejecting requests? I didn't test this, if you did and it works it's OK for me.

**Review by flash1293 — APPROVED** [review]
LGTM

**Pull request merged**